src/nvim/lua/treesitter.c

```diff
@@ -669,9 +677,85 @@ static int parser_get_timeout(lua_State *L)
   }
 
   lua_pushinteger(L, (long)ts_parser_timeout_micros(*p));
+  return 1;
+}
+
+typedef struct {
+  LuaRef cb;
+  lua_State *lstate;
+  bool lex;
+  bool parse;
+} TSLuaLoggerOpts;
+
+static void logger_cb(void *payload, TSLogType logtype, const char *s)
+{
+  TSLuaLoggerOpts *opts = (TSLuaLoggerOpts *)payload;
+  if ((!opts->lex && logtype == TSLogTypeLex)
+      || (!opts->parse && logtype == TSLogTypeParse)) {
+    return;
+  }
+
+  MAXSIZE_TEMP_ARRAY(args, 2);
+  ADD_C(args, STRING_OBJ(cstr_as_string(logtype == TSLogTypeParse ? "parse" : "lex")));
+  ADD_C(args, STRING_OBJ(cstr_as_string((char *)s)));
+  nlua_call_ref_with_lstate(opts->lstate, opts->cb, NULL, args, false, NULL);
+}
+
+static int parser_set_logger(lua_State *L)
```

Comment:
what is the high-level docstring for this? Is the logger per-parser?

should this be quasi-private? Once we have a logging API maybe we won't need this?

Reply:
>  Is the logger per-parser?

Yes, this is implied by the function prefix and the fact it is contained in `parser_meta`.

> should this be quasi-private? Once we have a logging API maybe we won't need this?

Yeah maybe. This file is just dumb bindings into treesitter and the *real* API is managed in Lua. I'll prefix it with an underscore.